File: src/client/python/agentguard/u_guard/remote_client.py
"""Remote guard client: talk to the server decision service over HTTP."""
from __future__ import annotations

import threading
import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass
from typing import Any

from agentguard.schemas.context import RuntimeContext
from agentguard.schemas.decisions import GuardDecision
from agentguard.schemas.events import RuntimeEvent
from agentguard.utils.errors import RemoteGuardError
from agentguard.utils.json import safe_dumps, safe_loads


@dataclass
class CircuitBreaker:
    """Simple open/closed breaker based on consecutive failures."""

    threshold: int = 3
    reset_after_s: float = 15.0
    _failures: int = 0
    _opened_at: float = 0.0

    @property
    def is_open(self) -> bool:
        if self._failures < self.threshold:
            return False
        if (time.time() - self._opened_at) > self.reset_after_s:
            # Half-open: allow a trial request.
            self._failures = self.threshold - 1
            return False
        return True

    def record_success(self) -> None:
        self._failures = 0
        self._opened_at = 0.0

    def record_failure(self) -> None:
        self._failures += 1
        if self._failures >= self.threshold:
            self._opened_at = time.time()


class RemoteGuardClient:
    def __init__(
        self,
        server_url: str | None,
        *,
        api_key: str | None = None,
        session_id: str | None = None,
        agent_id: str | None = None,
        user_id: str | None = None,
        session_key: str | None = None,
        user_ticket: str | None = None,
        timeout_s: float = 5.0,
        retries: int = 2,
        decide_path: str = "/v1/server/guard/decide",
        snapshot_path: str = "/v1/server/policy/snapshot",
        trace_path: str = "/v1/server/trace/upload",
        tool_report_path: str = "/v1/server/tools/report",
        mcp_report_path: str = "/v1/server/mcps/report",
        tool_sync_path: str = "/v1/server/tools/sync",
        approval_path: str = "/v1/server/approvals/{ticket_id}",
        register_path: str = "/v1/server/session/register",
        unregister_path: str = "/v1/server/session/unregister",
        agent_register_path: str = "/v1/server/agents/register",
        agent_sync_path: str = "/v1/server/agents/sync",
        runtime_session_create_path: str = "/v1/server/session/create",
        runtime_session_refresh_path: str = "/v1/server/session/refresh",
        runtime_session_close_path: str = "/v1/server/session/close",
        session_token: str | None = None,
        dpop_proof_factory: Any | None = None,
        use_dpop_auth: bool = False,
        legacy_identity_headers: bool = True,
        approval_wait_timeout_s: float = 600.0,
        approval_wait_chunk_s: float = 25.0,
    ) -> None:
        self.server_url = (server_url or "").rstrip("/")
        self.api_key = api_key
        self.session_id = session_id
        self.agent_id = agent_id
        self.user_id = user_id
        self.session_key = session_key
        self.user_ticket = user_ticket
        self.timeout_s = timeout_s
        self.retries = retries
        self.decide_path = decide_path
        self.snapshot_path = snapshot_path
        self.trace_path = trace_path
        self.tool_report_path = tool_report_path
        self.mcp_report_path = mcp_report_path
        self.tool_sync_path = tool_sync_path
        self.approval_path = approval_path
        self.register_path = register_path
        self.unregister_path = unregister_path
        self.agent_register_path = agent_register_path
        self.agent_sync_path = agent_sync_path
        self.runtime_session_create_path = runtime_session_create_path
        self.runtime_session_refresh_path = runtime_session_refresh_path
        self.runtime_session_close_path = runtime_session_close_path
        self.session_token = session_token
        self.dpop_proof_factory = dpop_proof_factory
        self.use_dpop_auth = use_dpop_auth
        self.legacy_identity_headers = legacy_identity_headers
        self.approval_wait_timeout_s = approval_wait_timeout_s
        self.approval_wait_chunk_s = max(1.0, approval_wait_chunk_s)
        self.breaker = CircuitBreaker()

    @property
    def enabled(self) -> bool:
        return bool(self.server_url)

    def _require_runtime_auth(self, purpose: str) -> None:
        if not self.use_dpop_auth or not self.session_token or not callable(self.dpop_proof_factory):
            raise RemoteGuardError(
                f"{purpose} requires a runtime-auth session_token and DPoP proof factory; create a runtime session first"
            )

    # ---- public API ----------------------------------------------------
    def decide(
        self,
        event: RuntimeEvent,
        context: RuntimeContext,
        *,
        trajectory_window: list[RuntimeEvent] | None = None,
        local_signals: list[str] | None = None,
        extensions: dict[str, Any] | None = None,
        client_cached_entries: list[dict[str, Any]] | None = None,
    ) -> GuardDecision:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        self._require_runtime_auth("remote guard decisions")
        if self.breaker.is_open:
            raise RemoteGuardError("circuit breaker open")

        body = {
            "request_id": f"req_{event.event_id}",
            "current_event": event.to_dict(),
            "context": context.to_dict(),
            "trajectory_window": [e.to_dict() for e in (trajectory_window or [])],
            "local_signals": list(local_signals or event.risk_signals),
            "policy_version": context.policy_version,
            "extensions": extensions or {},
            "client_cached_entries": list(client_cached_entries or []),
        }
        payload = self._post(self.decide_path, body)
        decision = payload.get("decision") or {}
        if not decision:
            raise RemoteGuardError("server returned no decision")
        gd = GuardDecision.from_dict(decision)
        for s in payload.get("risk_signals") or []:
            if s not in gd.risk_signals:
                gd.risk_signals.append(s)
        gd.metadata.setdefault("plugin_result", payload.get("plugin_result") or {})
        gd.metadata.setdefault("source", "remote")
        gd = self._await_review_resolution(gd)
        return gd

    def fetch_snapshot(self) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        self._require_runtime_auth("remote policy snapshot fetches")
        return self._get(self.snapshot_path)

    def upload_trace(self, trace: dict[str, Any]) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        self._require_runtime_auth("remote trace uploads")
        return self._post(self.trace_path, trace)

    def report_tool(self, context: RuntimeContext, tool: dict[str, Any]) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        self._require_runtime_auth("remote tool reports")
        body = {
            "context": context.to_dict(),
            "tool": tool,
        }
        return self._post(self.tool_report_path, body)

    def report_mcps(self, context: RuntimeContext, mcps: list[dict[str, Any]], scan: dict[str, Any] | None = None) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        self._require_runtime_auth("remote MCP reports")
        body = {
            "context": context.to_dict(),
            "mcps": list(mcps),
            "scan": scan if isinstance(scan, dict) else {},
        }
        return self._post(self.mcp_report_path, body)

    def sync_tools(self, context: RuntimeContext, tools: list[dict[str, Any]]) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        self._require_runtime_auth("remote tool sync")
        body = {
            "context": context.to_dict(),
            "tools": list(tools),
        }
        return self._post(self.tool_sync_path, body)

    def register_agent(self, agent: dict[str, Any]) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        return self._post(self.agent_register_path, dict(agent))

    def sync_agents(self, catalog: dict[str, Any]) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        return self._post(self.agent_sync_path, dict(catalog))

    def register_session(self, context: RuntimeContext) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        self._require_runtime_auth("runtime session sync")
        return self._post(self.register_path, {"context": context.to_dict()})

    def unregister_session(self) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        self._require_runtime_auth("runtime session cleanup")
        return self._post(self.unregister_path, {})

    def create_runtime_session(
        self,
        body: dict[str, Any],
        *,
        extra_headers_factory: Any | None = None,
    ) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        return self._post(self.runtime_session_create_path, dict(body), extra_headers_factory=extra_headers_factory)

    def refresh_runtime_session(self) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        self._require_runtime_auth("runtime session refresh")
        return self._post(self.runtime_session_refresh_path, {})

    def close_runtime_session(self) -> dict[str, Any]:
        if not self.enabled:
            raise RemoteGuardError("no server_url configured")
        self._require_runtime_auth("runtime session close")
        return self._post(self.runtime_session_close_path, {})

    def upload_trace_async(
        self,
        trace: dict[str, Any],
        *,
        on_success: Any | None = None,
        on_error: Any | None = None,
    ) -> threading.Thread | None:
        if not self.enabled:
            return None

        def _worker() -> None:
            try:
                self.upload_trace(trace)
                if callable(on_success):
                    on_success()
            except Exception as exc:  # background sync should not affect agent flow
                if callable(on_error):
                    on_error(exc)

        thread = threading.Thread(target=_worker, daemon=True)
        thread.start()
        return thread

    # ---- transport -----------------------------------------------------
    def _headers(self, *, method: str | None = None, url: str | None = None) -> dict[str, str]:
        headers = {"Content-Type": "application/json", "Accept": "application/json"}
        if self.use_dpop_auth:
            if self.session_token:
                headers["Authorization"] = f"DPoP {self.session_token}"
            elif self.api_key:
                headers["Authorization"] = f"Bearer {self.api_key}"
            if callable(self.dpop_proof_factory) and method and url:
                headers["DPoP"] = str(self.dpop_proof_factory(method, url, self.session_token))
        elif self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        if self.session_id and self.legacy_identity_headers:
            headers["X-AgentGuard-Session-Id"] = self.session_id
        if self.agent_id and self.legacy_identity_headers:
            headers["X-AgentGuard-Agent-Id"] = self.agent_id
        if self.user_id and self.legacy_identity_headers:
            headers["X-AgentGuard-User-Id"] = self.user_id
        if self.session_key and self.legacy_identity_headers:
            headers["X-AgentGuard-Session-Key"] = self.session_key
        if self.user_ticket:
            headers["X-AgentGuard-User-Ticket"] = self.user_ticket
        return headers

    def _request(
        self,
        method: str,
        path: str,
        body: dict | None,
        *,
        extra_headers_factory: Any | None = None,
    ) -> dict[str, Any]:
        url = f"{self.server_url}{path}"
        data = safe_dumps(body).encode("utf-8") if body is not None else None
        last_exc: Exception | None = None
        for attempt in range(self.retries + 1):
            headers = self._headers(method=method, url=url)
            if callable(extra_headers_factory):
                headers.update(dict(extra_headers_factory(method, url, body or {}) or {}))
            req = urllib.request.Request(
                url,
                data=data,
                headers=headers,
                method=method,
            )
            try:
                with urllib.request.urlopen(req, timeout=self.timeout_s) as resp:
                    raw = resp.read().decode("utf-8")
                self.breaker.record_success()
                return safe_loads(raw, fallback={}) or {}
            except urllib.error.HTTPError as exc:
                body = ""
                try:
                    body = exc.read().decode("utf-8", errors="replace")
                except Exception:
                    body = ""
                last_exc = RemoteGuardError(
                    f"remote guard call failed: HTTP {exc.code} {exc.reason}"
                    + (f" body={body}" if body else "")
                )
                if 500 <= int(exc.code) < 600 and attempt < self.retries:
                    time.sleep(min(0.2 * (2**attempt), 1.0))
                    continue
                break
            except (urllib.error.URLError, TimeoutError, OSError) as exc:
                last_exc = exc
                if attempt < self.retries:
                    time.sleep(min(0.2 * (2**attempt), 1.0))
        self.breaker.record_failure()
        if isinstance(last_exc, RemoteGuardError):
            raise last_exc
        raise RemoteGuardError(f"remote guard call failed: {last_exc}")

    def _post(
        self,
        path: str,
        body: dict,
        *,
        extra_headers_factory: Any | None = None,
    ) -> dict[str, Any]:
        return self._request("POST", path, body, extra_headers_factory=extra_headers_factory)

    def _get(self, path: str) -> dict[str, Any]:
        return self._request("GET", path, None)

    def _await_review_resolution(self, decision: GuardDecision) -> GuardDecision:
        if not (decision.requires_user or decision.requires_remote):
            return decision
        ticket_id = str(
            decision.metadata.get("review_ticket_id")
            or decision.metadata.get("ticket_id")
            or ""
        ).strip()
        if not ticket_id:
            return decision

        timeout_s = self.approval_wait_timeout_s
        deadline = None if timeout_s <= 0 else (time.time() + timeout_s)
        max_wait_s = max(self.timeout_s - 0.5, 1.0)
        while True:
            remaining = None if deadline is None else (deadline - time.time())
            if remaining is not None and remaining <= 0:
                return decision
            wait_s = max_wait_s if remaining is None else min(max_wait_s, remaining)
            wait_s = min(wait_s, self.approval_wait_chunk_s)
            path = self.approval_path.format(
                ticket_id=urllib.parse.quote(ticket_id, safe="")
            )
            payload = self._get(f"{path}?wait_ms={int(max(wait_s, 0.0) * 1000)}")
            status = str(payload.get("status") or "").lower()
            if status in {"approved", "denied"}:
                resolved = payload.get("resolved_decision")
                if isinstance(resolved, dict) and resolved.get("decision_type"):
                    return GuardDecision.from_dict(resolved)
                return decision
            if status != "pending":
                return decision

File: src/client/js/agentguard/u_guard/remote_client.js
"use strict";

const { GuardDecision } = require("../schemas/decisions");
const { RemoteGuardError } = require("../utils/errors");

class CircuitBreaker {
  constructor({ threshold = 3, reset_after_s = 15.0 } = {}) {
    this.threshold = threshold;
    this.reset_after_s = reset_after_s;
    this.failures = 0;
    this.opened_at = 0;
  }

  get is_open() {
    if (this.failures < this.threshold) {
      return false;
    }
    if (Date.now() / 1000 - this.opened_at > this.reset_after_s) {
      this.failures = this.threshold - 1;
      return false;
    }
    return true;
  }

  record_success() {
    this.failures = 0;
    this.opened_at = 0;
  }

  record_failure() {
    this.failures += 1;
    if (this.failures >= this.threshold) {
      this.opened_at = Date.now() / 1000;
    }
  }
}

class RemoteGuardClient {
  constructor(server_url = null, options = {}) {
    this.server_url = (server_url || "").replace(/\/$/, "");
    this.api_key = options.api_key || options.apiKey || null;
    this.session_id = options.session_id || options.sessionId || null;
    this.agent_id = options.agent_id || options.agentId || null;
    this.user_id = options.user_id || options.userId || null;
    this.session_key = options.session_key || options.sessionKey || null;
    this.user_ticket = options.user_ticket || options.userTicket || null;
    this.timeout_s = options.timeout_s ?? options.timeoutS ?? 5.0;
    this.retries = options.retries ?? 2;
    this.decide_path = options.decide_path || "/v1/server/guard/decide";
    this.snapshot_path = options.snapshot_path || "/v1/server/policy/snapshot";
    this.trace_path = options.trace_path || "/v1/server/trace/upload";
    this.tool_report_path = options.tool_report_path || "/v1/server/tools/report";
    this.tool_sync_path = options.tool_sync_path || options.toolSyncPath || "/v1/server/tools/sync";
    this.skill_report_path = options.skill_report_path || options.skillReportPath || "/v1/server/skills/report";
    this.mcp_report_path = options.mcp_report_path || options.mcpReportPath || "/v1/server/mcps/report";
    this.agent_register_path = options.agent_register_path || options.agentRegisterPath || "/v1/server/agents/register";
    this.agent_bootstrap_path = options.agent_bootstrap_path || options.agentBootstrapPath || "/v1/server/agents/bootstrap";
    this.agent_sync_path = options.agent_sync_path || options.agentSyncPath || "/v1/server/agents/sync";
    this.runtime_session_create_path = options.runtime_session_create_path || options.runtimeSessionCreatePath || "/v1/server/session/create";
    this.runtime_session_refresh_path = options.runtime_session_refresh_path || options.runtimeSessionRefreshPath || "/v1/server/session/refresh";
    this.runtime_session_close_path = options.runtime_session_close_path || options.runtimeSessionClosePath || "/v1/server/session/close";
    this.runtime_plugin_config_path = options.runtime_plugin_config_path || options.runtimePluginConfigPath || "/v1/server/session/plugin-config";
    this.approval_path = options.approval_path || "/v1/server/approvals/{ticket_id}";
    this.register_path = options.register_path || "/v1/server/session/register";
    this.unregister_path = options.unregister_path || "/v1/server/session/unregister";
    this.session_token = options.session_token || options.sessionToken || null;
    this.dpop_proof_factory = options.dpop_proof_factory || options.dpopProofFactory || null;
    this.use_dpop_auth = Boolean(options.use_dpop_auth || options.useDpopAuth);
    this.legacy_identity_headers = options.legacy_identity_headers ?? options.legacyIdentityHeaders ?? true;
    this.approval_wait_timeout_s = options.approval_wait_timeout_s ?? options.approvalWaitTimeoutS ?? 600.0;
    this.approval_wait_chunk_s = Math.max(1.0, options.approval_wait_chunk_s ?? options.approvalWaitChunkS ?? 25.0);
    this.breaker = new CircuitBreaker();
  }

  get enabled() {
    return Boolean(this.server_url);
  }

  requireRuntimeAuth(purpose) {
    if (!this.use_dpop_auth || !this.session_token || typeof this.dpop_proof_factory !== "function") {
      throw new RemoteGuardError(
        `${purpose} requires a runtime-auth session_token and DPoP proof factory; create a runtime session first`,
      );
    }
  }

  async decide(event, context, options = {}) {
    if (!this.enabled) {
      throw new RemoteGuardError("no server_url configured");
    }
    this.requireRuntimeAuth("remote guard decisions");
    if (this.breaker.is_open) {
      throw new RemoteGuardError("circuit breaker open");
    }
    const payload = await this.post(this.decide_path, {
      request_id: `req_${event.event_id}`,
      current_event: event.toDict(),
      context: context.toDict(),
      trajectory_window: (options.trajectory_window || []).map((item) => item.toDict()),
      local_signals: options.local_signals || event.risk_signals || [],
      policy_version: context.policy_version,
      extensions: options.extensions || {},
      client_cached_entries: options.client_cached_entries || [],
    });
    const decision = GuardDecision.fromDict(payload.decision || {});
    for (const signal of payload.risk_signals || []) {
      if (!decision.risk_signals.includes(signal)) {
        decision.risk_signals.push(signal);
      }
    }
    decision.metadata.plugin_result = decision.metadata.plugin_result || payload.plugin_result || {};
    decision.metadata.source = decision.metadata.source || "remote";
    return this.awaitReviewResolution(decision);
  }

  fetch_snapshot() {
    this.requireRuntimeAuth("remote policy snapshot fetches");
    return this.get(this.snapshot_path);
  }

  upload_trace(trace) {
    this.requireRuntimeAuth("remote trace uploads");
    return this.post(this.trace_path, trace);
  }

  report_tool(context, tool) {
    this.requireRuntimeAuth("remote tool reports");
    return this.post(this.tool_report_path, {
      context: context.toDict(),
      tool,
    });
  }

  sync_tools(context, tools) {
    this.requireRuntimeAuth("remote tool sync");
    return this.post(this.tool_sync_path, {
      context: context.toDict(),
      tools: Array.isArray(tools) ? tools : [],
    });
  }

  report_skills(context, skills, scan = {}) {
    this.requireRuntimeAuth("remote skill reports");
    return this.post(this.skill_report_path, {
      context: context.toDict(),
      skills: Array.isArray(skills) ? skills : [],
      scan: scan && typeof scan === "object" && !Array.isArray(scan) ? scan : {},
    });
  }

  report_mcps(context, mcps, scan = {}) {
    this.requireRuntimeAuth("remote MCP reports");
    return this.post(this.mcp_report_path, {
      context: context.toDict(),
      mcps: Array.isArray(mcps) ? mcps : [],
      scan: scan && typeof scan === "object" && !Array.isArray(scan) ? scan : {},
    });
  }

  register_session(context) {
    this.requireRuntimeAuth("runtime session sync");
    return this.post(this.register_path, {
      context: context.toDict(),
    });
  }

  unregister_session() {
    this.requireRuntimeAuth("runtime session cleanup");
    return this.post(this.unregister_path, {});
  }

  register_agent(agent) {
    return this.post(this.agent_register_path, { ...(agent || {}) });
  }

  bootstrap_agents(catalog) {
    return this.post(this.agent_bootstrap_path, { ...(catalog || {}) });
  }

  sync_agents(catalog) {
    return this.post(this.agent_sync_path, { ...(catalog || {}) });
  }

  create_runtime_session(body, { extra_headers_factory = null, extraHeadersFactory = extra_headers_factory } = {}) {
    return this.post(this.runtime_session_create_path, { ...(body || {}) }, { extra_headers_factory: extraHeadersFactory });
  }

  refresh_runtime_session() {
    this.requireRuntimeAuth("runtime session refresh");
    return this.post(this.runtime_session_refresh_path, {});
  }

  fetch_runtime_plugin_config() {
    this.requireRuntimeAuth("runtime plugin-config fetches");
    return this.get(this.runtime_plugin_config_path);
  }

  close_runtime_session() {
    this.requireRuntimeAuth("runtime session close");
    return this.post(this.runtime_session_close_path, {});
  }

  upload_trace_async(trace, { on_success = null, on_error = null } = {}) {
    return this.upload_trace(trace).then(() => {
      if (typeof on_success === "function") {
        on_success();
      }
    }).catch((error) => {
      if (typeof on_error === "function") {
        on_error(error);
      }
    });
  }

  headers({ method = null, url = null } = {}) {
    const headers = {
      "Content-Type": "application/json",
      Accept: "application/json",
    };
    if (this.use_dpop_auth) {
      if (this.session_token) {
        headers.Authorization = `DPoP ${this.session_token}`;
      } else if (this.api_key) {
        headers.Authorization = `Bearer ${this.api_key}`;
      }
      if (typeof this.dpop_proof_factory === "function" && method && url) {
        headers.DPoP = String(this.dpop_proof_factory(method, url, this.session_token));
      }
    } else if (this.api_key) {
      headers.Authorization = `Bearer ${this.api_key}`;
    }
    if (this.session_id && this.legacy_identity_headers) {
      headers["X-AgentGuard-Session-Id"] = this.session_id;
    }
    if (this.agent_id && this.legacy_identity_headers) {
      headers["X-AgentGuard-Agent-Id"] = this.agent_id;
    }
    if (this.user_id && this.legacy_identity_headers) {
      headers["X-AgentGuard-User-Id"] = this.user_id;
    }
    if (this.session_key && this.legacy_identity_headers) {
      headers["X-AgentGuard-Session-Key"] = this.session_key;
    }
    if (this.user_ticket && !this.session_token) {
      headers["X-AgentGuard-User-Ticket"] = this.user_ticket;
    }
    return headers;
  }

  async request(method, path, body = null, { extra_headers_factory = null } = {}) {
    const url = `${this.server_url}${path}`;
    let lastError = null;
    for (let attempt = 0; attempt <= this.retries; attempt += 1) {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), this.timeout_s * 1000);
      const headers = this.headers({ method, url });
      if (typeof extra_headers_factory === "function") {
        Object.assign(headers, extra_headers_factory(method, url, body || {}) || {});
      }
      try {
        const response = await fetch(url, {
          method,
          headers,
          body: body == null ? undefined : JSON.stringify(body),
          signal: controller.signal,
        });
        clearTimeout(timeout);
        if (!response.ok) {
          let detail = "";
          try {
            const text = await response.text();
            if (text) {
              const parsed = JSON.parse(text);
              detail = parsed && parsed.detail ? String(parsed.detail) : text;
            }
          } catch (_) {
            detail = "";
          }
          throw new Error(`HTTP ${response.status}${detail ? `: ${detail}` : ""}`);
        }
        this.breaker.record_success();
        return await response.json();
      } catch (error) {
        clearTimeout(timeout);
        lastError = error;
      }
    }
    this.breaker.record_failure();
    throw new RemoteGuardError(`remote guard call failed: ${String(lastError && lastError.message ? lastError.message : lastError)}`);
  }

  post(path, body, options = {}) {
    return this.request("POST", path, body, options);
  }

  get(path) {
    return this.request("GET", path, null);
  }

  async awaitReviewResolution(decision) {
    if (!(decision.requires_user || decision.requires_remote)) {
      return decision;
    }
    const ticketId = String(
      decision.metadata.review_ticket_id || decision.metadata.ticket_id || "",
    ).trim();
    if (!ticketId) {
      return decision;
    }
    const timeoutS = Number(this.approval_wait_timeout_s || 0);
    const deadline = timeoutS > 0 ? Date.now() + timeoutS * 1000 : null;
    const maxWaitS = Math.max(Number(this.timeout_s || 0) - 0.5, 1.0);
    while (true) {
      const remainingMs = deadline == null ? null : deadline - Date.now();
      if (remainingMs != null && remainingMs <= 0) {
        return decision;
      }
      const waitS = remainingMs == null
        ? Math.min(this.approval_wait_chunk_s, maxWaitS)
        : Math.min(this.approval_wait_chunk_s, maxWaitS, remainingMs / 1000);
      const path = this.approval_path.replace(
        "{ticket_id}",
        encodeURIComponent(ticketId),
      );
      const payload = await this.get(`${path}?wait_ms=${Math.max(0, Math.floor(waitS * 1000))}`);
      const status = String(payload && payload.status ? payload.status : "").toLowerCase();
      if (status === "approved" || status === "denied") {
        const resolved = payload && payload.resolved_decision;
        if (resolved && typeof resolved === "object" && resolved.decision_type) {
          return GuardDecision.fromDict(resolved);
        }
        return decision;
      }
      if (status !== "pending") {
        return decision;
      }
    }
  }
}

module.exports = {
  CircuitBreaker,
  RemoteGuardClient,
};
